Validation/HGCalValidation/plugins/CaloParticleValidation.cc

```diff
@@ -0,0 +1,267 @@
+                                  Histograms_CaloParticleValidation&) const override;
+
+      void dqmAnalyze(edm::Event const&,
+                              edm::EventSetup const&,
+                              Histograms_CaloParticleValidation const&) const override;
+
+      // ----------member data ---------------------------
+      std::string folder_;
+      std::vector<int> particles_to_monitor_;
+
+      edm::EDGetTokenT<std::vector<SimVertex> > simVertices_;
+      edm::EDGetTokenT<std::vector<CaloParticle> > caloParticles_;
+      edm::EDGetTokenT<std::vector<reco::SuperCluster>> simPFClusters_;
+      edm::EDGetTokenT<reco::PFCandidateCollection> simPFCandidates_;
+      edm::EDGetTokenT<HGCRecHitCollection> recHitsEE_;
+      edm::EDGetTokenT<HGCRecHitCollection> recHitsFH_;
+      edm::EDGetTokenT<HGCRecHitCollection> recHitsBH_;
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+CaloParticleValidation::CaloParticleValidation(const edm::ParameterSet& iConfig)
+  : folder_(iConfig.getParameter<std::string>("folder")),
+  particles_to_monitor_(iConfig.getParameter<std::vector<int> >("particles_to_monitor")),
+  simVertices_(consumes<std::vector<SimVertex>>(iConfig.getParameter<edm::InputTag>("simVertices"))),
+  caloParticles_(consumes<std::vector<CaloParticle> >(iConfig.getParameter<edm::InputTag>("caloParticles"))),
+  simPFClusters_(consumes<std::vector<reco::SuperCluster>>(iConfig.getParameter<edm::InputTag>("simPFClusters"))),
+  simPFCandidates_(consumes<reco::PFCandidateCollection>(iConfig.getParameter<edm::InputTag>("simPFCandidates"))),
+  recHitsEE_(consumes<HGCRecHitCollection>(iConfig.getParameter<edm::InputTag>("recHitsEE"))),
+  recHitsFH_(consumes<HGCRecHitCollection>(iConfig.getParameter<edm::InputTag>("recHitsFH"))),
+  recHitsBH_(consumes<HGCRecHitCollection>(iConfig.getParameter<edm::InputTag>("recHitsBH")))
+{
+   //now do what ever initialization is needed
+}
+
+
+CaloParticleValidation::~CaloParticleValidation()
+{
+   // do anything here that needs to be done at desctruction time
+   // (e.g. close files, deallocate resources etc.)
+}
+
+
+//
+// member functions
+//
+
+// ------------ method called for each event  ------------
+
+void
+CaloParticleValidation::dqmAnalyze(edm::Event const& iEvent, edm::EventSetup const& iSetup,
+                      Histograms_CaloParticleValidation const & histos) const
+{
+  using namespace edm;
+
+  Handle<HGCRecHitCollection> recHitHandleEE;
+  Handle<HGCRecHitCollection> recHitHandleFH;
+  Handle<HGCRecHitCollection> recHitHandleBH;
+  // make a map detid-rechit
+
+  iEvent.getByToken(recHitsEE_, recHitHandleEE);
+  iEvent.getByToken(recHitsFH_, recHitHandleFH);
+  iEvent.getByToken(recHitsBH_, recHitHandleBH);
+  const auto& rechitsEE = *recHitHandleEE;
+  const auto& rechitsFH = *recHitHandleFH;
+  const auto& rechitsBH = *recHitHandleBH;
+  std::map<DetId, const HGCRecHit*> hitmap;
```

Comment:
maybe faster as `unordered_map`

Reply:
Should also bring in hashing for DetId: I doubt this will be a performance bottleneck for the upgrade.....

Reply:
See #22537, if it gets merged first maybe it can be incorporated here